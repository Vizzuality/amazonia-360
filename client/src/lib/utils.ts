import { type ClassValue, clsx } from "clsx";
import { twMerge } from "tailwind-merge";

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs));
}

export const getKeys = Object.keys as <T extends object>(
  obj: T,
) => Array<keyof T>;

export function getContrastColor(hexColor: string): number {
  // Convert hex color to RGB
  const r: number = parseInt(hexColor.substring(1, 3), 16);
  const g: number = parseInt(hexColor.substring(3, 5), 16);
  const b: number = parseInt(hexColor.substring(5, 7), 16);

  // Calculate luminance (brightness) of the color
  const luminance: number = (0.299 * r + 0.587 * g + 0.114 * b) / 255;

  // Return black or white depending on luminance
  return luminance;
}

export function joinWithAnd(arr: string[]) {
  if (arr.length === 0) return "";
  if (arr.length === 1) return arr[0];
  if (arr.length === 2) return arr.join(" and ");

  const lastItem = arr.pop();
  return arr.join(", ") + " and " + lastItem;
}
