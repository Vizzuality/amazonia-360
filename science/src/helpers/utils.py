import json

import geopandas as gpd
from shapely.geometry import Polygon


def transform_polygon_to_rings(geojson):
    """
    Transform a GeoJSON polygon into a format compatible with the ArcGIS API.

    Parameters:
    geojson (dict): A GeoJSON dictionary.

    Returns:
    str: A JSON string with the polygon coordinates.
    """
    poly = Polygon(geojson["geometry"]["coordinates"][0])
    gdf = gpd.GeoDataFrame(index=[0], crs="epsg:4326", geometry=[poly])
    gdf = gdf.to_crs(epsg=102100)
    poly = gdf.geometry[0]
    coordinates = list(poly.exterior.coords)
    # Transform coordinates into the required format
    rings = [[list(coord) for coord in coordinates]]
    poly_dict = {"rings": rings}
    poly_json = json.dumps(poly_dict)
    return poly_json
