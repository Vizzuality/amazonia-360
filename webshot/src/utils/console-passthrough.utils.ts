import { ConsoleMessage } from "playwright";
import { inspect } from "util";

export const consolePassthrough = (msg: ConsoleMessage) => {
  for (let i = 0; i < msg.args().length; ++i) {
    msg
      .args()
      [i].jsonValue()
      .then((value) => console.debug(inspect(value, undefined, 5)))
      .catch((err) => console.debug(`Failed to get value: ${err}`));
  }
};
