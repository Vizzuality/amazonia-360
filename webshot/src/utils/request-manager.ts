import { Logger } from "@nestjs/common";
import { debounce } from "lodash";
import { Page } from "playwright";

export const createRequestManager = (
  page: Page,
  logger: Logger,
  timeoutMs: number = 30_000,
  debounceDelayMs: number = 10_000,
) => {
  const activeRequests: Set<string> = new Set();
  let timeoutId: NodeJS.Timeout | undefined = undefined;

  let resolve: (() => void) | undefined = undefined;
  let reject: (() => void) | undefined = undefined;
  const promise = new Promise<void>((_resolve, _reject) => {
    resolve = _resolve;
    reject = _reject;
  });

  // This function is debounced so that every time a request is started, finished or failed, a
  // small delay is created before we check again if the network is idle
  const checkIfIdle = debounce(() => {
    if (activeRequests.size === 0) {
      if (timeoutId !== undefined) {
        clearTimeout(timeoutId);
      }
      logger.log("Network is idle");
      resolve?.();
    }
  }, debounceDelayMs);

  return {
    setupEvents: () => {
      page.on("request", (request) => {
        const url = request.url();
        // blob requests are ignored as they may be linked to Mapbox WebWorkers and the front-end
        // application doesn't create any
        if (!url.startsWith("blob:")) {
          activeRequests.add(url);
          checkIfIdle();
        }
      });

      page.on("requestfinished", (request) => {
        const url = request.url();
        activeRequests.delete(url);
        checkIfIdle();
      });

      page.on("requestfailed", (request) => {
        const url = request.url();
        activeRequests.delete(url);
        checkIfIdle();
      });
    },
    isIdle: () => {
      // Reject if the timeout has passed
      timeoutId = setTimeout(() => {
        reject?.();
      }, timeoutMs);

      return promise;
    },
  };
};
