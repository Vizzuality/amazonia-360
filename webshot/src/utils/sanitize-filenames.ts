export const sanitizeFilename = (filename: string): string => {
  return filename.replace(/[^A-Za-z0-9\.\-_]/g, "");
};
