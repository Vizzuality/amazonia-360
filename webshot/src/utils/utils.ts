/**
 * Namespace containing utility functions for value validation.
 */
export class Value {
  /**
   * Checks if a value is neither null nor undefined.
   *
   * @template T - The type of the value being checked
   * @param a - The value to check
   * @returns A type predicate indicating if the value is valid (not null or undefined)
   *
   * @example
   * ```typescript
   * const value: string | null = getSomeValue();
   * if (Value.isValid(value)) {
   *   // value is typed as string here
   *   console.log(value.length);
   * }
   * ```
   */
  static isValid = <T>(a: T | null | undefined): a is T => a !== undefined && a !== null;
}
