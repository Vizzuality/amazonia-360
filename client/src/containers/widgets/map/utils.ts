export const FALLBACK_WIDGET_DEFAULT_BASEMAP_ID: BasemapIds = "gray-vector";
import { omit } from "@/lib/utils";

import { Indicator } from "@/app/local-api/indicators/route";
import {
  IndicatorView,
  DefaultTopicConfig,
  DefaultTopicIndicatorConfig,
  TopicView,
} from "@/app/parsers";

import { BasemapIds } from "@/components/map/controls/basemap";

type MapIndicatorProperties = "basemapId" | "opacity";

export const handleMapIndicatorPropertyChange = (
  propertyName: MapIndicatorProperties,
  propertyValue: BasemapIds | number,
  overviewTopicsData: DefaultTopicConfig[] | null,
  indicator: Indicator,
  setSyncDefaultTopics: (
    callback: (prevDefaultTopics: DefaultTopicConfig[] | null) => DefaultTopicConfig[] | null,
  ) => void,
  setTopics: (callback: (prevTopicsState: TopicView[] | null) => TopicView[] | null) => void,
  defaultValues: { basemapId: BasemapIds; opacity: number },
) => {
  const currentIndicatorTopicId = indicator.topic.id;
  const defaultValue = defaultValues[propertyName];
  const isResettingToDefault = propertyValue === defaultValue;

  const isManagingViaSyncDefaultTopics = overviewTopicsData?.some(
    (topic) => currentIndicatorTopicId === topic.id,
  );

  if (isManagingViaSyncDefaultTopics) {
    setSyncDefaultTopics(
      (prevDefaultTopics: DefaultTopicConfig[] | null): DefaultTopicConfig[] | null => {
        const newTopicsArray: DefaultTopicConfig[] = prevDefaultTopics
          ? [...prevDefaultTopics]
          : [];
        const topicIndex = newTopicsArray.findIndex((t) => t.id === currentIndicatorTopicId);

        if (topicIndex !== -1) {
          const targetTopic = { ...newTopicsArray[topicIndex] };
          const indicators: DefaultTopicIndicatorConfig[] = targetTopic.indicators
            ? [...targetTopic.indicators]
            : [];
          const indicatorConfigIndex = indicators.findIndex((i) => i.id === indicator.id);

          if (isResettingToDefault) {
            if (indicatorConfigIndex !== -1) {
              const indicatorConfig = { ...indicators[indicatorConfigIndex] };
              delete indicatorConfig[propertyName];

              // eslint-disable-next-line @typescript-eslint/no-unused-vars
              const { id, ...overrides } = indicatorConfig;
              if (Object.keys(overrides).length > 0) {
                indicators[indicatorConfigIndex] = indicatorConfig;
              } else {
                indicators.splice(indicatorConfigIndex, 1);
              }
            }
          } else {
            if (indicatorConfigIndex !== -1) {
              indicators[indicatorConfigIndex] = {
                ...indicators[indicatorConfigIndex],
                [propertyName]: propertyValue,
              };
            } else {
              indicators.push({
                id: indicator.id,
                [propertyName]: propertyValue,
              } as DefaultTopicIndicatorConfig);
            }
          }

          if (indicators.length === 0) {
            newTopicsArray.splice(topicIndex, 1);
          } else {
            newTopicsArray[topicIndex] = {
              ...targetTopic,
              indicators,
            };
          }
        } else if (!isResettingToDefault) {
          newTopicsArray.push({
            id: currentIndicatorTopicId,
            indicators: [
              {
                id: indicator.id,
                [propertyName]: propertyValue,
              } as DefaultTopicIndicatorConfig,
            ],
          });
        }

        return newTopicsArray.length > 0 ? newTopicsArray : null;
      },
    );
  } else {
    setTopics((prevTopicsState: TopicView[] | null): TopicView[] | null => {
      const currentTopics = prevTopicsState || [];
      const topicIndex = currentTopics.findIndex((t) => t.id === indicator.topic.id);

      if (topicIndex === -1) {
        return prevTopicsState;
      }

      const topicToUpdate = currentTopics[topicIndex];
      const indicatorExists = (topicToUpdate.indicators || []).some(
        (ind: IndicatorView) => ind.id === indicator.id && ind.type === "map",
      );

      if (!indicatorExists) {
        return prevTopicsState;
      }

      const newTopics = currentTopics.map((topic, index) => {
        if (index === topicIndex) {
          return {
            ...topic,
            indicators: (topic.indicators || []).map((ind) => {
              if (ind.id === indicator.id && ind.type === "map") {
                if (isResettingToDefault) {
                  return omit(ind, [propertyName]);
                } else {
                  return { ...ind, [propertyName]: propertyValue };
                }
              }
              return ind;
            }),
          };
        }
        return topic;
      });

      return newTopics.length > 0 ? newTopics : null;
    });
  }
};
