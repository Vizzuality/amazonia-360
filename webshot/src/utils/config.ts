import config from "config";
import { Value } from "./utils";

/**
 * Config class to get configuration values from the config package
 */
export class Config {
  /**
   * Get a configuration value
   *
   * @param key The configuration key
   * @param defaultValue The default value if the key is not found
   * @returns The configuration value or the default value if the key is not found
   *
   * @throws Error if the key is not found and no default value is provided
   */
  public static get<T>(key: string, defaultValue?: T): T {
    if (config.has(key)) {
      const value = config.get<T>(key);

      if (Value.isValid(value)) {
        return value;
      } else if (Value.isValid(defaultValue)) {
        return defaultValue;
      }
    } else if (Value.isValid(defaultValue)) {
      return defaultValue;
    }

    throw new Error(
      `Config key ${key} is not defined and no default value is provided`,
    );
  }

  /**
   * Get a string configuration value
   *
   * @param key The configuration key
   * @param defaultValue The default value if the key is not found
   * @returns The configuration value or the default value if the key is not found
   *
   * @throws Error if the key is not found and no default value is provided
   * @throws Error if the configuration value is not a string
   */
  public static getString(key: string, defaultValue?: string): string {
    if (config.has(key)) {
      const value = config.get<string>(key);

      if (typeof value === "string") {
        return value;
      }

      throw new Error(`Config key ${key} is not a string`);
    } else if (Value.isValid(defaultValue)) {
      return defaultValue;
    }

    throw new Error(
      `Config key ${key} is not defined and no default value is provided`,
    );
  }

  /**
   * Get a boolean configuration value
   *
   * @param key The configuration key
   * @param defaultValue The default value if the key is not found
   * @returns The configuration value or the default value if the key is not found
   *
   * @throws Error if the key is not found and no default value is provided
   * @throws Error if the configuration value is not a boolean
   */
  public static getBoolean(key: string, defaultValue?: boolean): boolean {
    if (config.has(key)) {
      const value = config.get<boolean>(key);

      if (typeof value === "boolean") {
        return value;
      }

      throw new Error(`Config key ${key} is not a boolean`);
    } else if (Value.isValid(defaultValue)) {
      return defaultValue;
    }

    throw new Error(
      `Config key ${key} is not defined and no default value is provided`,
    );
  }

  /**
   * Get a number configuration value
   *
   * @param key The configuration key
   * @param defaultValue The default value if the key is not found
   * @returns The configuration value or the default value if the key is not found
   *
   * @throws Error if the key is not found and no default value is provided
   * @throws Error if the configuration value is not a number
   */
  public static getNumber(key: string, defaultValue?: number): number {
    if (config.has(key)) {
      const value = config.get<number>(key);

      if (typeof value === "number") {
        return value;
      } else if (typeof Number(value) === "number" && !isNaN(Number(value))) {
        return Number(value);
      }

      throw new Error(`Config key ${key} is not a number`);
    } else if (Value.isValid(defaultValue)) {
      return defaultValue;
    }

    throw new Error(
      `Config key ${key} is not defined and no default value is provided`,
    );
  }

  /**
   * Retrieves a configuration value as either a string array or a single string.
   *
   * This method looks for a string value under the provided key and processes it as follows:
   * - If the value contains commas, it splits the string and returns the resulting array.
   * - If the value has no commas, it returns the single string value.
   *
   * @param {string} key - The configuration key to retrieve
   * @param {string[] | string} [defaultValue] - The default value to return if the key isn't found
   * @returns {string[] | string} An array of strings if commas are present, otherwise a single string
   *
   * @throws {Error} If the key doesn't exist and no default value is provided
   * @throws {Error} If the value exists but is not a string
   *
   * @example
   * // If config has "allowedOrigins=http://localhost:3000,https://example.com"
   * const origins = Config.getStringAsArray("allowedOrigins");
   * // returns ["http://localhost:3000", "https://example.com"]
   *
   * @example
   * // If config has "defaultLanguage=en"
   * const language = Config.getStringAsArray("defaultLanguage");
   * // returns "en"
   */
  public static getStringAsArray(
    key: string,
    defaultValue?: string[] | string,
  ): string[] | string {
    if (config.has(key)) {
      const value = config.get<string>(key);

      if (typeof value === "string") {
        const values = value.split(",");
        return values.length > 1 ? values : values[0];
      }

      throw new Error(`Config key ${key} is not a string`);
    } else if (Value.isValid(defaultValue)) {
      return defaultValue;
    }

    throw new Error(
      `Config key ${key} is not defined and no default value is provided`,
    );
  }

  /**
   * Retrieves and merges values from two config sources into a single array.
   *
   * This method looks for values in two places:
   * 1. An array stored under the provided `arrayKey`
   * 2. A comma-separated string stored under the provided `stringKey`
   *
   * It combines all values found in either source into a single array with duplicates removed.
   * If neither source contains values, returns the provided default value.
   *
   * @template T - The type of elements in the array
   * @param {string} arrayKey - The config key to look for an array of values
   * @param {string} stringKey - The config key to look for a comma-separated string of values
   * @param {T[]} [defaultValue=[]] - Default value to return if no values are found
   * @returns {T[]} An array containing all unique values from both sources, or the default value if no values were found
   *
   * @remarks
   * When retrieving values from the string source, each value is cast to type T using `as unknown as T`.
   * This may cause type safety issues if the string values can't be properly converted to type T.
   */
  public static getMergedArray<T>(
    arrayKey: string,
    stringKey: string,
    defaultValue: T[] = [],
  ): T[] {
    const values: Set<T> = new Set<T>();

    if (config.has(arrayKey)) {
      const value = config.get<T[]>(arrayKey);

      if (Array.isArray(value)) {
        value.forEach((value) => {
          values.add(value);
        });
      } else {
        console.warn(`Config ${arrayKey} is not an array`);
      }
    } else {
      console.warn(`Config ${arrayKey} does not exist`);
    }

    if (config.has(stringKey)) {
      const value = config.get<string>(stringKey);

      if (typeof value === "string") {
        value.split(",").forEach((value) => {
          values.add(value as unknown as T);
        });
      } else {
        console.warn(`Config ${stringKey} is not a string`);
      }
    } else {
      console.warn(`Config ${stringKey} does not exist`);
    }

    return values.size > 0 ? Array.from(values) : defaultValue;
  }
}
