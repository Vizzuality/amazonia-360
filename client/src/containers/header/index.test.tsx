import Header from "@/containers/header";

import { render, screen } from "@/jest.utils";

const mockUsePathname = jest.fn<string, unknown[]>();
jest.mock("next/navigation", () => ({
  usePathname() {
    return mockUsePathname();
  },
}));

describe("Header", () => {
  beforeEach(() => {
    mockUsePathname.mockReturnValue("/");
  });

  it("renders Header component without crashing", () => {
    render(<Header />);
    expect(screen.getByRole("banner")).toBeInTheDocument();
  });

  it("renders the title", () => {
    render(<Header />);
    expect(screen.getByText("Amazonia360")).toBeInTheDocument();
  });

  it("renders the logo", () => {
    render(<Header />);
    expect(screen.getByAltText("IDB")).toBeInTheDocument();
  });

  it("renders the badge", () => {
    render(<Header />);
    expect(screen.getByText("Prototype")).toBeInTheDocument();
  });

  // Selected
  it("renders the Home link selected", () => {
    const { container } = render(<Header />);
    expect(container).toMatchSnapshot();
  });

  it("renders the Report Tool link selected", () => {
    mockUsePathname.mockReturnValue("/report");
    const { container } = render(<Header />);
    expect(container).toMatchSnapshot();
  });

  it("renders the Hub link selected", () => {
    mockUsePathname.mockReturnValue("/hub");
    const { container } = render(<Header />);
    expect(container).toMatchSnapshot();
  });
});
