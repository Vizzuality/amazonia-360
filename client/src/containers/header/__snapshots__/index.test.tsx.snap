// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders the Home link selected 1`] = `
<div>
  <header
    class="flex h-20 flex-col justify-center bg-white backdrop-blur"v
  >
    <div
      class="mx-4 flex items-center justify-between md:container md:mx-auto"
    >
      <div
        class="z-[120] flex items-center space-x-2"
      >
        <a
          href="/"
        >
          <h1
            class="flex items-center space-x-4"
          >
            <img
              alt="IDB"
              data-nimg="1"
              decoding="async"
              height="24"
              loading="lazy"
              src="/IDB-logo.svg"
              style="color: transparent;"
              width="65"
            />
            <div
              class="space-x-2"
            >
              <span
                class="text-sm font-medium text-blue-500"
              >
                AmazoniaForever360+
              </span>
            </div>
          </h1>
        </a>
        <button
          aria-controls="radix-:rc:"
          aria-expanded="false"
          aria-haspopup="dialog"
          data-state="closed"
          type="button"
        >
          <div
            class="inline-flex items-center rounded-md border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 border-transparent bg-secondary text-secondary-foreground hover:bg-secondary/80"
          >
            Beta
          </div>
        </button>
      </div>
      <nav
        class="flex space-x-8 print:hidden"
      >
        <a
          class="text-sm hover:text-cyan-500 text-cyan-500"
          href="/"
        >
          Home
        </a>
        <a
          class="text-sm hover:text-cyan-500"
          href="/report"
        >
          Report Tool
        </a>
        <a
          class="text-sm hover:text-cyan-500"
          href="/hub"
        >
          Hub
        </a>
      </nav>
    </div>
  </header>
</div>
`;

exports[`Header renders the Hub link selected 1`] = `
<div>
  <header
    class="flex h-20 flex-col justify-center bg-white backdrop-blur" 
  >
    <div
      class="mx-4 flex items-center justify-between md:container md:mx-auto"
    >
      <div
        class="z-[120] flex items-center space-x-2"
      >
        <a
          href="/"
        >
          <h1
            class="flex items-center space-x-4"
          >
            <img
              alt="IDB"
              data-nimg="1"
              decoding="async"
              height="24"
              loading="lazy"
              src="/IDB-logo.svg"
              style="color: transparent;"
              width="65"
            />
            <div
              class="space-x-2"
            >
              <span
                class="text-sm font-medium text-blue-500"
              >
                AmazoniaForever360+
              </span>
            </div>
          </h1>
        </a>
        <button
          aria-controls="radix-:ri:"
          aria-expanded="false"
          aria-haspopup="dialog"
          data-state="closed"
          type="button"
        >
          <div
            class="inline-flex items-center rounded-md border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 border-transparent bg-secondary text-secondary-foreground hover:bg-secondary/80"
          >
            Beta
          </div>
        </button>
      </div>
      <nav
        class="flex space-x-8 print:hidden"
      >
        <a
          class="text-sm hover:text-cyan-500"
          href="/"
        >
          Home
        </a>
        <a
          class="text-sm hover:text-cyan-500"
          href="/report"
        >
          Report Tool
        </a>
        <a
          class="text-sm hover:text-cyan-500 text-cyan-500"
          href="/hub"
        >
          Hub
        </a>
      </nav>
    </div>
  </header>
</div>
`;

exports[`Header renders the Report Tool link selected 1`] = `
<div>
  <header
    class="flex h-20 flex-col justify-center bg-white backdrop-blur" 
  >
    <div
      class="mx-4 flex items-center justify-between md:container md:mx-auto"
    >
      <div
        class="z-[120] flex items-center space-x-2"
      >
        <a
          href="/"
        >
          <h1
            class="flex items-center space-x-4"
          >
            <img
              alt="IDB"
              data-nimg="1"
              decoding="async"
              height="24"
              loading="lazy"
              src="/IDB-logo.svg"
              style="color: transparent;"
              width="65"
            />
            <div
              class="space-x-2"
            >
              <span
                class="text-sm font-medium text-blue-500"
              >
                AmazoniaForever360+
              </span>
            </div>
          </h1>
        </a>
        <button
          aria-controls="radix-:rf:"
          aria-expanded="false"
          aria-haspopup="dialog"
          data-state="closed"
          type="button"
        >
          <div
            class="inline-flex items-center rounded-md border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 border-transparent bg-secondary text-secondary-foreground hover:bg-secondary/80"
          >
            Beta
          </div>
        </button>
      </div>
      <nav
        class="flex space-x-8 print:hidden"
      >
        <a
          class="text-sm hover:text-cyan-500"
          href="/"
        >
          Home
        </a>
        <a
          class="text-sm hover:text-cyan-500 text-cyan-500"
          href="/report"
        >
          Report Tool
        </a>
        <a
          class="text-sm hover:text-cyan-500"
          href="/hub"
        >
          Hub
        </a>
      </nav>
    </div>
  </header>
</div>
`;
